### PR TITLE
Compare boundary points instead of rows and cols when checking region containment

### PR DESCRIPTION
**Pull request by alvarotroya**
TLDR version: Add methods to compare the position of points and compare regions based on their boundaries.

Long version: I was thinking about python support and saw that the Region object holds position information that could potentially be used to get the scope of a node in python. I realized comparing regions is the same as comparing their boundary points, so I added methods to compare points and to get start and end points from regions. I had to remove the dependency of Point on Region, in order to avoid circular dependencies. I believe that thinking about the higher-level object 'Points' and 'Regions' might prove useful in the future as compared to thinking about rows and cols numbers :)

### CODE REVIEW

File: lua/refactoring/point.lua [hunk 1]
```diff
@@ -64,8 +58,41 @@ function Point:clone()
...
 end
 
-function Point:within_node(node)
-    return Region:from_node(node, 0):contains_point(self) -- buffer doesn't matter in this case
+--- Compare the position of two points.
+-- Given two points with positions pos_1 and pos_2
+-- return -1 if pos_1 < pos_2, 0 if pos_1 == pos_2 and 1 otherwise.
+-- @params point   the second point to compare to.
+-- @return int     either -1, 0 or 1
+function Point:compareTo(point)
```
Comment (ThePrimeagen) [inline lua/refactoring/point.lua:66]: could you project this to 1d

```
local function project_point(point)
    return point.row * 10000000 + point.col
end
```

That way we don't have to have a complicated `compareTo` function

Then also rename `compareTo` to `compare_to` for lua conventions.

Also we need some tests because `0 != false` in lua.  Therefore you have some functions that are incorrect.  `not 0 == false`.  Which means you have to do `... == 0` comparisons.

File: lua/refactoring/lsp_utils.lua [hunk 2]
```diff
@@ -16,7 +17,7 @@ end
 
 local function to_region(pointOrRegion)
     if not pointOrRegion.end_row then
-        return pointOrRegion:to_region()
+        return Region:from_point(pointOrRegion)
```
Comment (ThePrimeagen) [inline lua/refactoring/lsp_utils.lua:20]: lets rename `from_point` to `to_region` because its a region or a point, not a just a point.

File: lua/refactoring/region.lua [hunk 3]
```diff
@@ -72,6 +70,23 @@ function Region:from_node(node, bufnr)
     }, self)
 end
 
+--- Get a region from a given point.
+---@param point    the point to use as start- and endpoint
+---@param {bufnr}  the bufnr for the region
+---@return region
+function Region:from_point(point, bufnr)
```
Comment (ThePrimeagen) [inline lua/refactoring/region.lua:77]: its fine keeping this, its just incorrectly used above.

**Comment by ThePrimeagen** [comment]
I do like the comparison though btw.

File: lua/refactoring/point.lua (hunk 1 above)
Comment (alvarotroya) [reply lua/refactoring/point.lua:66]: Good point about the renaming! Will take care of that, I also just realised there is a simpler way of writing the function, will push that soon so that you can take a look. I personally think is much easier to think about simple if-statement comparisons but it seems you are convinced of the 1d-projection idea. Is there a reason for this? Would it be a bad idea to just keep the if-statement comparisons?

Comment (alvarotroya) [reply lua/refactoring/point.lua:66]: I don't think I am making any 0 == false comparisons in the code but it's a good point and I certainly didn't know that 0 is not falsy in lua. I will definitely write some tests and update the PR.

File: lua/refactoring/lsp_utils.lua (hunk 2 above)
Comment (alvarotroya) [reply lua/refactoring/lsp_utils.lua:20]: I am not sure what you mean here, the caller function is already named `to_region`. I suggest just using Region:from_values here, there is no other place where a point needs to be turned into a region (maybe we should try to adapt the caller so that we don't need to check here whether the arg is a point or a region in the first place)

Comment (alvarotroya) [reply lua/refactoring/lsp_utils.lua:20]: Actually I just checked the code, the call stack is 
```debug/printf.lua -> lsp_utils.insert_new_line_text -> to_region``` or
```lsp_utils_spec.lua -> lsp_utils.insert_test -> to_region```

However, both calls already pass a region. So at the moment the check is unnecessary. I think the easiest is just to have the callers passing regions and never points (there is an easy way of turning points into regions so this won't block future functionality). At the moment there are no other callers of this methods, so the decision can still be taken.

Comment (alvarotroya) [reply lua/refactoring/lsp_utils.lua:20]: Using Region:from_values at this place didn't add any value compared to the current state, it just added unnecessary extra lines

File: lua/refactoring/point.lua (hunk 1 above)
Comment (ThePrimeagen) [reply lua/refactoring/point.lua]: Tests great for all the new helper functions .  I think it will make sure that we don't have any oopsies and we already have some tests so it should be super simple to create them.

On projection point of view.  The reason why its nicer is that it turns a 5 branch function into a 3 branch.  Its just simpler code to look at.  at this point I am not strong enough against it, so whatever.

File: lua/refactoring/lsp_utils.lua (hunk 2 above)
Comment (ThePrimeagen) [reply lua/refactoring/lsp_utils.lua:20]: this seems fine then.

There is this small problem where a region exists within a text editor.  The problem becomes that currently a point doesn't, meaning there is no bufnr associated with a Point.  Something to think about for future me.

File: lua/refactoring/point.lua (hunk 1 above)
Comment (alvarotroya) [reply lua/refactoring/point.lua]: Thank you for your flexibility, Prime! I have added the tests and fixed the merge conflicts, I think this should be okay now :) let me know what you think

**Comment by ThePrimeagen** [comment]
I have always said I am grateful for any contributions!  Thanks for taking the time, its appreciated.

**Comment by alvarotroya** [comment]
> I have always said I am grateful for any contributions!  Thanks for taking the time, its appreciated. 

Thank you for making it possible! it was a pleasure to contribute even such a small thing to the project, looking forward to seeing it grow 😊
